core/object/thread_runner.cpp

```diff
@@ -0,0 +1,207 @@
+/*************************************************************************/
+/*  thread_runner.cpp                                                    */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "thread_runner.h"
+
+#include "core/os/os.h"
+
+ThreadRunner *ThreadRunner::singleton = nullptr;
+
+void ThreadRunner::_process_task() {
+	task_mutex.lock();
+	Task *task = tasks_waiting.first()->self();
+	tasks_waiting.remove(tasks_waiting.first());
+	task_mutex.unlock();
+
+	if (task->native_func) {
+		task->native_func(task->native_func_userdata);
+	} else {
+		Callable::CallError ce;
+		Variant ret;
+		task->callable.call(nullptr, 0, ret, ce);
+	}
+
+	task->completed = true;
+	task->done_semaphore.post();
+}
+
+void ThreadRunner::_thread_function(void *p_user) {
```

Comment:
Concretely, maybe this object is an internal object only, and the user-visible object is a shim that is assigned to one of N such internal pools.   On a thread restricted platform, all shims would map to the same pool (N=1, with the number of threads set to the platform limit.)  On a large platform, you might allocate several internal pools if you want to always allocate reserved threads to different functional areas.   Not sure you need that, as the OS will likely do a decent job keeping things fair if you just put a lot of threads in the pool on a large platform.

Reply:
Game folks might like the idea though of saying "3 threads for this type of task, 7 threads for this type of task."

Reply:
On a big PC we can just preallocate as many threads as we want, but this API needs to ensure we can work with a small amount for platforms where this is troublesome (HTML5 and some consoles).

Reply:
Yes, so is this a singleton then?  Will it be provided in the namespace by default?

It would be nicer API if users can TaskRunner.new() and just get a shim that calls the single pool that Godot provides.  Don't make the user responsible for only instantiating one of these and have code from different modules/developers share it.